iroh-relay/src/server.rs

```diff
@@ -438,19 +478,50 @@ impl Server {
 async fn relay_supervisor(
     mut tasks: JoinSet<Result<()>>,
     mut relay_http_server: Option<http_server::Server>,
+    mut quic_server: Option<QuicServer>,
 ) -> Result<()> {
-    let res = match (relay_http_server.as_mut(), tasks.len()) {
-        (None, 0) => Ok(Err(anyhow!("Nothing to supervise"))),
-        (None, _) => tasks.join_next().await.expect("checked"),
-        (Some(relay), 0) => relay.task_handle().await.map(anyhow::Ok),
-        (Some(relay), _) => {
+    let res = match (
+        relay_http_server.as_mut(),
+        quic_server.as_mut(),
+        tasks.len(),
+    ) {
+        (None, None, 0) => Ok(Err(anyhow!("Nothing to supervise"))),
+        (None, None, _) => tasks.join_next().await.expect("checked"),
+        (None, Some(quic_ep), 0) => quic_ep.task_handle().await.map(anyhow::Ok),
+        (None, Some(quic_ep), _) => {
```

Comment:
This match explosion here is a bit unpleasant I guess.  I had to try this a few times but I think this might work to remove the match and do just a single `select!` statement for this entire supervisor:

```rust
    let quic_enabled = quic_server.is_some();
    let mut quic_fut = match quic_server {
        Some(ref mut server) => futures_util::future::Either::Left(server.task_handle()),
        None => futures_util::future::Either::Right(futures_lite::future::pending()),
    };
    tokio::select! {
        biased;
        ret = tasks.join_next(), if tasks.len() > 0 => todo!(),
        ret = &mut quic_fut, if quic_enabled => todo!(),
    }
```

The explanation:

- You can disable branches in the select using the `if`.
- A disabled branch won't poll the future from that branch.
- But even if disabled all futures are constructed so need to be valid.
- So you need a future that exists regardless of whether a server is present.
- Because of borrowing rules you need to do the `.is_some()` stuff up front.

Reply:
damn, nice.